### PR TITLE
Determine the build order when building the dependencies in Cartfile.resolved

### LINKED ISSUE #717: Sort Cartfile.resolved
Any operation that outputs Cartfile.resolved should save it using a line based sort.

This would really help with git diffs and merging. Currently the order of the file seems to be pretty arbitrary and random.

### PR DESCRIPTION
**Pull request by ikesyo**
~~The build order is not preserved in `Cartfile.resolved` now.~~

Trying to fix #717 based on the discussion: https://github.com/Carthage/Carthage/issues/717#issuecomment-135501479

### CODE REVIEW

File: Source/CarthageKit/Cartfile.swift [hunk 1]
```diff
@@ -270,14 +270,18 @@ public struct Dependency<V: VersionType>: Equatable {
 	/// The version(s) that are required to satisfy this dependency.
 	public var version: V
 
-	public init(project: ProjectIdentifier, version: V) {
+	/// The dependencies of this dependency.
+	public let dependencies: Set<ProjectIdentifier>
```
Comment (ikesyo) [inline Source/CarthageKit/Cartfile.swift]: This is intended to be used for https://github.com/Carthage/Carthage/pull/972#issuecomment-162215783.

**Comment by mdiep** [comment]
I think we need to break this into 2 changes, separated by several months and releases, in order to keep `Cartfile.resolved`s compatible with recent versions of Carthage.

1. Don't rely on the order of `Cartfile.resolved`: calculate dependencies when needed.
2. Output `Cartfile.resolved` in lexicographical order.

We need to allow a significant period of time between (1) and (2) so that everyone can upgrade to a version that does (1) before we release a version that does (2). Otherwise we run into an Xcode-upgrade type scenario.

File: Source/CarthageKitTests/ResolverSpec.swift [hunk 2]
```diff
@@ -84,11 +84,11 @@ class ResolverSpec: QuickSpec {
 
 			var generator = dependencies.generate()
 
-			// Dependencies should be listed in build order.
+			// Dependencies should be listed in lexical order of its project name.
```
Comment (mdiep) [inline Source/CarthageKitTests/ResolverSpec.swift]: We should make sure we add new tests that the dependencies in `Cartfile.resolved` are sorted in build order.

Comment (ikesyo) [inline Source/CarthageKitTests/ResolverSpec.swift]: Exactly. :bow:

**Comment by ikesyo** [comment]
Thanks for the idea, that seems totally reasonable. :smiley: 

> 1. Don't rely on the order of `Cartfile.resolved`: calculate dependencies when needed.

I would update this PR to work on only the thing.

**Comment by ikesyo** [comment]
It's ready. :zap:

**Comment by mdiep** [comment]
Thanks for tackling this! :sparkling_heart:

**Comment by schwa** [comment]
Thanks for this!

**Pull request merged**